PORT-7257 | Add the Upbound EKS as a service use case to our docs (#1221)
port-labs/port-docs: docs/create-self-service-experiences/setup-backend/github-workflow/examples/Upbound/upbound-port-capabilities.md

                "Denied": "red"
              }
            }
          },
          "required": []
        },
        "mirrorProperties": {},
        "calculationProperties": {},
        "relations": {
          "eks_cluster": {
            "title": "EKS Cluster",
            "description": "The cluster created for this request",
            "target": "eks_cluster",
            "required": false,
            "many": false
          },
          "upbound_control_plane": {
            "title": "Upbound Control Plane",
            "description": "The control plane this cluster was requested for",
            "target": "upbound_control_plane",
            "required": false,
            "many": false
          }
        }
      }
```
</details>
You will need to create these blueprints in your Port organization. 

:::warn Blueprint creation
Blueprint creation may fail if they are not created in the order that they appear above
:::

### Creating actions
Below are action blueprints which define the different actions we need, to trigger the different Github workflows.

These actions are defined on their appropriate blueprints.

::note Customizing GitHub values

> Remember to change `CHANGE_TO_YOUR_GITHUB_ORG_NAME` and `CHANGE_TO_YOUR_REPO_NAME` values in the action blueprint.  You need to replace these with your appropriate Github organization name and repository name.

Reviewer: ```suggestion
Remember to change `CHANGE_TO_YOUR_GITHUB_ORG_NAME` and `CHANGE_TO_YOUR_REPO_NAME` values in the action blueprint. You need to replace these with your appropriate Github organization name and repository name.
```

:::

<details>
<summary><b>Approve Cluster Request (click to expand)</b></summary>

```json showLineNumbers
{
  "identifier": "eks_cluster_request_approve_cluster_request",
  "title": "Approve Cluster Request",
  "icon": "GithubActions",
  "trigger": {
    "type": "self-service",
    "operation": "DAY-2",
    "userInputs": {
      "properties": {},
      "required": []
    },
    "blueprintIdentifier": "eks_cluster_request"
  },
  "invocationMethod": {
    "type": "GITHUB",
    "org": "CHANGE_TO_YOUR_GITHUB_ORG_NAME",
    "repo": "CHANGE_TO_YOUR_REPO_NAME",
    "workflow": "approve-cluster-request.yaml",
    "workflowInputs": {

>       "{{if (.inputs | has(\"ref\")) then \"ref\" else null end}}": "{{.inputs.\"ref\"}}",

Reviewer: Update the inputs syntax here to just take the input fields, there's no need for the `{{if (.inputs.....}}` in the input key selection

      "port_payload": {
        "action": "{{ .action.identifier[(\"eks_cluster_request_\" | length):] }}",
        "resourceType": "run",
        "status": "TRIGGERED",
        "trigger": "{{ .trigger | {by, origin, at} }}",

>         "context": {

Reviewer: This input does not appear in the workflow, it is also called context instead of being called port_context

          "entity": "{{.entity.identifier}}",
          "blueprint": "{{.action.blueprint}}",
          "runId": "{{.run.id}}"
        },
        "payload": {
          "entity": "{{ (if .entity == {} then null else .entity end) }}",
          "action": {
            "invocationMethod": {
              "type": "GITHUB",
              "omitPayload": false,
              "omitUserInputs": true,
              "reportWorkflowStatus": true,
              "org": "CHANGE_TO_YOUR_GITHUB_ORG_NAME",
              "repo": "CHANGE_TO_YOUR_REPO_NAME",
              "workflow": "approve-cluster-request.yaml"
            },
            "trigger": "{{.trigger.operation}}"
          },
          "properties": {},
          "censoredProperties": "{{.action.encryptedProperties}}"
        }
      }
    },
    "reportWorkflowStatus": true
  },
  "requiredApproval": false,
  "publish": true
}
```
</details>

<details>
<summary><b>Deny Cluster Request (click to expand)</b></summary>

```json showLineNumbers

> {

Reviewer: Same comments as previous Port SSA

        },
        "node-size": {
          "title": "Node Size",
          "type": "string",
          "default": "small",
          "enum": [
            "small",
            "medium",
            "large"
          ],
          "enumColors": {
            "small": "lightGray",
            "medium": "lightGray",
            "large": "lightGray"
          }
        },
        "node-count": {
          "icon": "DefaultProperty",
          "title": "Node Count",
          "type": "string",
          "default": "1"
        },
        "control-plane": {
          "icon": "DefaultProperty",
          "title": "Upbound control plane",
          "type": "string",
          "blueprint": "upbound_control_plane",
          "format": "entity"
        }
      },
      "required": [
        "cluster-name",
        "control-plane"
      ],
      "order": [
        "control-plane",
        "cluster-name",
        "node-size",
        "node-count"
      ]
    },
    "blueprintIdentifier": "eks_cluster"
  },
  "invocationMethod": {
    "type": "GITHUB",
    "org": "CHANGE_TO_YOUR_GITHUB_ORG_NAME",
    "repo": "CHANGE_TO_YOUR_REPO_NAME",
    "workflow": "new-cluster-request.yaml",

>     "workflowInputs": {

Reviewer: Same comments as previous Port SSA

          "type": "string"
        },
        "node-size": {
          "title": "Node Size",
          "type": "string",
          "default": "small",
          "enum": [
            "small",
            "medium",
            "large"
          ],
          "enumColors": {
            "small": "lightGray",
            "medium": "lightGray",
            "large": "lightGray"
          }
        },
        "node-count": {
          "icon": "DefaultProperty",
          "title": "Node Count",
          "type": "string",
          "default": "1"
        },
        "control-plane": {
          "title": "Upbound control plane",
          "type": "string",
          "blueprint": "upbound_control_plane",
          "format": "entity"
        }
      },
      "required": [
        "cluster-name",
        "control-plane"
      ],
      "order": [
        "control-plane",
        "cluster-name",
        "node-size",
        "node-count"
      ]
    },
    "blueprintIdentifier": "eks_cluster"
  },
  "invocationMethod": {
    "type": "GITHUB",
    "org": "CHANGE_TO_YOUR_GITHUB_ORG_NAME",
    "repo": "CHANGE_TO_YOUR_REPO_NAME",
    "workflow": "new-cluster-request.yaml",

>     "workflowInputs": {

Reviewer: Same comments as previous Port SSA

            "{{if (.inputs | has(\"node-count\")) then \"node-count\" else null end}}": "{{.inputs.\"node-count\"}}",
            "{{if (.inputs | has(\"control-plane\")) then \"control-plane\" else null end}}": "{{.inputs.\"control-plane\" | if type == \"array\" then map(.identifier) else .identifier end}}"
          },
          "censoredProperties": "{{.action.encryptedProperties}}"
        }
      }
    },
    "reportWorkflowStatus": true
  },
  "requiredApproval": false,
  "publish": true
}
```

</details>

<details>
<summary><b>Delete EKS Cluster (click to expand)</b></summary>

```json showLineNumbers
{
  "identifier": "eks_cluster_delete_eks_cluster",
  "title": "Delete EKS Cluster",
  "icon": "Alert",
  "trigger": {
    "type": "self-service",
    "operation": "DELETE",
    "userInputs": {
      "properties": {},
      "required": [],
      "order": []
    },
    "blueprintIdentifier": "eks_cluster"
  },
  "invocationMethod": {
    "type": "GITHUB",
    "org": "CHANGE_TO_YOUR_GITHUB_ORG_NAME",
    "repo": "CHANGE_TO_YOUR_REPO_NAME",
    "workflow": "delete-cluster.yaml",

>     "workflowInputs": {

Reviewer: Same comments as previous Port SSA

        "status": "TRIGGERED",
        "trigger": "{{ .trigger | {by, origin, at} }}",
        "context": {
          "entity": "{{.entity.identifier}}",
          "blueprint": "{{.action.blueprint}}",
          "runId": "{{.run.id}}"
        },
        "payload": {
          "entity": "{{ (if .entity == {} then null else .entity end) }}",
          "action": {
            "invocationMethod": {
              "type": "GITHUB",
              "omitPayload": false,
              "omitUserInputs": true,
              "reportWorkflowStatus": true,
              "org": "CHANGE_TO_YOUR_GITHUB_ORG_NAME",
              "repo": "CHANGE_TO_YOUR_REPO_NAME",
              "workflow": "delete-cluster.yaml"
            },
            "trigger": "{{.trigger.operation}}"
          },
          "properties": {},
          "censoredProperties": "{{.action.encryptedProperties}}"
        }
      }
    },
    "reportWorkflowStatus": true
  },
  "requiredApproval": true,
  "approvalNotification": {
    "type": "email"
  },
  "publish": true
}
```

</details>

### Creating Upbound control plane Port entities
After setting up the Port blueprints and actions, we need to insert some entities manually.

> These entities will represent the different Upbound control planes which were created earlier.

Reviewer: ```suggestion
These entities will represent the different Upbound control planes which you created in your Upbound organization.
```

To do this, follow these steps:

1. Navigate to the [Upbound control planes](https://app.getport.io/upbound_control_planes) catalog page.

2. Click the `Manually add Upbound Control Plane` button (or the `+ Upbound Control plane` at the top right of the page).
<img src='/img/create-self-service-experiences/setup-backend/github-workflow/examples/Upbound/addUpboundControlPlaneManually.png' border='1px' />

> 3. In the `identifier` field, insert the Upbound control plane `identifier` which we saved earlier and click create (do this step multiple times if there are more than 1 control planes).

Reviewer: ```suggestion
3. In the `identifier` field, insert the Upbound control plane `identifier` which we saved earlier and click create (do this step multiple times if there is more than one control plane).
```

<img src='/img/create-self-service-experiences/setup-backend/github-workflow/examples/Upbound/setUpboundControlPlaneIdentifier.png' border='1px' />

## Using Port

> At this point, everything should be set up. Browse to your [Self-service](https://app.getport.io/self-serve) page to view the different actions you defined in Port, and try them out.

Reviewer: Add an explanation about the example flows that are available using these actions:

- Create an EKS cluster from Port (as a Port admin/platform engineer)
- Request an EKS cluster from Port (as a Port member/developer)
- Approve an EKS cluster request
- Deny an EKS cluster request
- Delete an EKS cluster
